User can specify key manticore options which are loaded with sensible defaults

kokrui opened the pull request:
As per https://github.com/trailofbits/MUI/issues/29

Adds a small UI change that allows user to quickly alter important manticore options
![image](https://user-images.githubusercontent.com/29654756/148781624-395e3de1-8aae-48c7-8667-c4714afaee0d.png)

Options presented based on the run-dialog in binja plugin.
Validation is handled by manticore itself, where if an option is invalid the stack trace is printed to the MUI log.

kokrui committed:
emulate binja rundialog settings

kokrui committed:
fix setting data structure

kokrui committed:
implement new MUISetup UI, except array/bool

kokrui committed:
cleanup setup panel

kokrui committed:
run mcore with specified args

kokrui committed:
lint

sschriner commented:
Closes https://github.com/trailofbits/MUI/issues/29

kokrui committed:
missed out 1 setting name

kokrui committed:
ci set fetch-depth 0

ekilmer commented on MUI/src/main/java/mui/MUIRunner.java:
```diff
@@ -23,11 +27,33 @@ public void stopProc() {
+			String optType =
+				(String) MUISettings.SETTINGS.get("NATIVE_RUN_SETTINGS").get(name)[0].get("type");
+
+			if (optType == "string" || optType == "number") {
+
+			}
+			else if (optType == "array") {
+				//TODO: implement
+			}
```
Maybe throw some real warning messages or even exceptions?

ekilmer commented on MUI/src/main/java/mui/MUIRunner.java:
```diff
@@ -23,11 +27,33 @@ public void stopProc() {
 	}
 
-	public void callProc(String[] manticoreArgs) {
+	public void constructCommand(String manticoreExePath,
```
This doesn't seem to be used anywhere? Looks like `buildFormPanel` has some similar logic

kokrui replied:
!! forgot to remove that

ekilmer commented on MUI/src/main/java/mui/MUISetupProvider.java:
```diff
@@ -43,61 +53,110 @@ private void setLogProvider(MUILogProvider log) {
-		mainPanelConstraints.fill = GridBagConstraints.BOTH;
-		mainPanelConstraints.gridwidth = GridBagConstraints.REMAINDER;
-		mainPanelConstraints.weightx = 0.9;
-		mainPanelConstraints.weighty = 0.9;
+				});
 
-		inputPanel = new JPanel(new GridBagLayout());
+				inputRowConstraints.gridx = 3;
+				inputRowConstraints.gridwidth = 1;
+				inputRowConstraints.weightx = 0.25;
+				inputRow.add(selectButton, inputRowConstraints);
 
-		inputPanelConstraints = new GridBagConstraints();
-		inputPanelConstraints.fill = GridBagConstraints.BOTH;
+				formPanel.add(inputRow);
 
-		inputPanelConstraints.gridx = 0;
-		inputPanelConstraints.gridy = 0;
-		inputPanelConstraints.weightx = 0.25;
-		inputPanelConstraints.gridwidth = 1;
-		inputPanel.add(new JLabel("Program Path:"), inputPanelConstraints);
+			}
+			else if (prop.get("type") == "string" || prop.get("type") == "number") {
+				entry.setText(prop.get("default").toString());
+				entry.setToolTipText("Only 1 value allowed");
+				formPanel.add(entry);
+			}
+			else if (prop.get("type") == "array") {
+				// TODO: doesn't handle default param for arrays, but not needed as part of sensible defaults for running manticore on native binaries
+				// for now, same UI as string/num, and we will parse space-separated args
+				entry.setText(prop.get("default").toString());
+				entry.setToolTipText("You can space-separate multiple arguments");
+				formPanel.add(entry);
+			}
+			else {
+				// TODO: to achieve parity with Binja MUI, type==boolean must be supported, but not needed as part of sensible defaults for running manticore on native binaries
+				throw new UnsupportedOperationException(
+					String.format("[ERROR] Cannot create input row for %s with the type %s", name,
+						prop.get("type")));
+			}
```
Would probably be easier to read if you created functions that modified `entry` for each of the respective property types

ekilmer reviewed (changes requested):
Looking good! I forget if the Binja GUI plugin has this, but it would be nice to have a text box for "extra Manticore arguments", similar to the original screen but also in addition to some sensible default values and options.

Only some minor nitpicks to address. Once those are done, I think it'll be okay to merge!

kokrui committed:
remove unused function

kokrui committed:
separate input UIs into functions

kokrui committed:
allow set additional mcore args

kokrui commented:
resolved the issues and added the "extra mcore args" field!

ekilmer committed:
Remove spotless code format ratcheting

We don't have that many files right now, and ratcheting requires
fetching origin/main in CI, which can slow things down.

If the formatting pass becomes a bottleneck, we can turn this back on

ekilmer reviewed (approved):
LGTM!
